### PR TITLE
created cloud function for fetching friends' schedules securely

### PR DESCRIPTION
**Pull request by yatharth-b**
Resolves gt-scheduler/website#164

This PR creates a new Cloud Function to securely fetch the schedules of a user's friends given a term, friends' UID, and schedule version IDs.

Request Payload:

```
{
    "IDToken": string // token id provided by frontend auth,
    "friends": {
        _FRIEND_ID_: [list of version ids]
    },
    "term": string
}
```

Response Payload
```
{
    _FRIEND_ID_: {
        "versions": {
            _VERSION_ID_: {
                "name": string,
                "schedule": schedule object
            }
        }
    }
}
```

For testing purposes, comment out code from the function that verifies the user's IdToken.

### CODE REVIEW

File: functions/src/fetch_friend_versions.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+      // Authenticate token id
+      auth
+        .verifyIdToken(IDToken)
+        .then(async (decodedToken: admin.auth.DecodedIdToken) => {
+          const uid = decodedToken.uid;
+          // Fetch user's friends
+          let userFriendData : FriendsData = await friendsCollection
+            .doc(uid)
+            .get()
+            .then((doc) => doc.data()) as FriendsData;
```
Comment (nhatnghiho) [inline functions/src/fetch_friend_versions.ts:43]: If we're already providing the friend's uid and the needed versions, why do we need to access `friendsCollection` again? Can't we just use the provided uid and versions to fetch schedules?

File: functions/src/fetch_friend_versions.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+const auth = admin.auth();
+const schedulesCollection = admin.firestore().collection("schedules-dev");
+
+const corsHandler = cors({ origin: true });
+
+interface FriendsData {
+  terms: Record<string, Record<string, string[]>>[];
+}
+
+type VersionReturn = Record<string, string[]>;
```
Comment (nhatnghiho) [inline functions/src/fetch_friend_versions.ts:16]: I haven't been able to use the endpoint due to the proposed changes, but I'm a bit confused as to why this is only `Record<string, string[]>`. Aren't we fetching a schedule of the following type?

image

**Review by nhatnghiho — CHANGES_REQUESTED** [review]

File: functions/src/fetch_friend_versions.ts (hunk 1 above)
Comment (yatharth-b) [reply functions/src/fetch_friend_versions.ts:43]: that is to verify that the request body does not include any friends' schedules that the user does not have access to.

File: functions/src/fetch_friend_versions.ts (hunk 2 above)
Comment (yatharth-b) [reply functions/src/fetch_friend_versions.ts]: yea, my bad

**Comment by yatharth-b** [comment]
need to test more once we make an api call from the frontend

**Review by nhatnghiho — APPROVED** [review]

**Pull request merged**